### ISSUE TITLE
Web console support

### ISSUE THREAD

**Issue by ziajka** [GNS3/gns3-web-ui#64]
https://xtermjs.org/

**Comment by ziajka** [comment]
https://github.com/paradoxxxzero/butterfly

**Comment by piotrpekala7** [comment]
We can simply add reg file:

Windows Registry Editor Version 5.00
[HKEY_CLASSES_ROOT\putty]
    @="URL: Putty Protocol handler"
    "URL Protocol"=""
	"CustomUrlApplication"="C:\\Program Files\\PuTTY\\putty.exe"
	"CustomUrlArguments"="\"%1\""
    [HKEY_CLASSES_ROOT\putty\shell]
    [HKEY_CLASSES_ROOT\putty\shell\open]
    [HKEY_CLASSES_ROOT\putty\shell\open\command]
    @="\"C:\\Program Files\\PuTTY\\putty.exe\" %1"

then in script file: reg import .\PuttyProtocolHandler.reg
unfortunately putty won't parse as telnet app does
**When using a custom URL, the entire URL is passed to the registered app/script as a single parameter. Your app/script needs to parse the URL to extract what it needs.**

**Comment by piotrpekala7** [comment]
btw. that looks pretty cool https://support.shotgunsoftware.com/hc/en-us/articles/219031308-Launching-applications-using-custom-browser-protocols

**Comment by grossmj** [comment]
This is exactly what EVE-NG does, they use protocol handlers and thus require their users to install a client side pack. They have one pack for Windows (http://eve-ng.net/downloads/windows-client-side-pack) and one for Linux

Scripts for Windows: https://github.com/dainok/unetlab/tree/master/windows
Scripts for Linux: https://github.com/SmartFinn/eve-ng-integration for Linux

**Comment by candlerb** [comment]
I can't see how to connect to console inside gns3 web UI.  Did this feature make it into gns3-server v2.2.0?  And if so, how do you use it?  (Right-click on node doesn't give me a console option)

Aside 1: I made a workaround using shellinabox, see [this gist](https://gist.github.com/candlerb/12607ee118248de9ac70a3e77b9f2b78).

Aside 2: I would have asked on community.gns3.com, but it is currently down due to "Invalid SSL certificate" (error from CloudFlare).

**Comment by piotrpekala7** [comment]
@candlerb Fixed with https://github.com/GNS3/gns3-web-ui/issues/515. It should work properly with newest version (we're planning release in 2 weeks). User would have possibility to start console by clicking context menu on link. Currently there is option to achieve that by selecting only one link between nodes and then right-click

**Comment by candlerb** [comment]
Thanks!  I can't get anything to happen by click on the link between nodes, but happy to wait for next version.

**Comment by grossmj** [comment]
@candlerb for something to happen, you will have to install our webclient pack https://github.com/GNS3/gns3-webclient-pack (in beta for now). This will install protocol handlers so you can use local terminal applications to console to nodes in the web-ui

**Comment by candlerb** [comment]
I understand, thanks.

The application here is for students in a lab being able to access the consoles on their virtual routers/switches in a single, shared emulation.  We can get them to install a telnet client such as putty on their own laptops, but they don't need the full-blown GNS3 client (the instructor has that).

Previously we used dynamips + dynagen, and this gave static console ports, so we could just give a handout saying which port to connect to.

With GNS3 the ports are dynamic ([discussion](https://community.gns3.com/community/discussion/possible-to-assign-fixed-tcp-por)) which makes it a bit trickier.

One solution would be to use the API to dynamically generate a webpage of devices and ports, and the students just read off the right port to connect to.  I can build that easily enough.  If the emulation is restarted, all the ports may change.

The shellinabox workaround I posted above is better, as it keeps everything inside the browser and needs no telnet client at all. It seems OK, but hasn't been tested under classroom load yet.

In an ideal world:

* console exposed via web/websocket based frontend built into GNS3, e.g. `/v2/projects/X/nodes/Y/console`
* students connect to the GNS3 web interface to view the topology and select the device to connect to
* only the instructor can modify the topology, but students can open consoles and start/stop devices

This is quite a specialized requirement though, so I think coding against the API is a reasonable way to approach this.

Anyway, thanks again for the info, and for your ongoing development work.  GNS3 is turning into a very capable product, and I just thought you might like to know how we're planning to use it :-)
